### ISSUE TITLE
GLXBadContext in 1.0.3

### ISSUE THREAD

**Issue by dmd** [pauldmccarthy/fsleyes#76]
One of our users is getting:

```
$ fsleyes -V
fsleyes/FSLeyes version 1.0.3

$ fsleyes
The program 'fsleyes' received an X Window System error.
This probably reflects a bug in the program.
The error was 'GLXBadContext'.
  (Details: serial 7541 error_code 161 request_code 149 minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the --sync command line
   option to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

On their local machine (MacOS) I verified:

```
$ defaults read org.macosforge.xquartz.X11 enable_iglx
1

$ defaults read org.xquartz.X11 enable_iglx
1
```

and they've rebooted since writing those defaults. They are using XQuartz 2.8.1.

Any suggestions for how we can debug this?

**Comment by pauldmccarthy** [comment]
Hi @dmd, does `glxgears` / `glxinfo` work for your user?

**Comment by dmd** [comment]
He writes:

>>>

```
[dharper@micc ~]$ glxgears
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
19127 frames in 9.5 seconds = 2002.991 FPS
516 frames in 8.7 seconds = 59.487 FPS
706 frames in 8.6 seconds = 81.784 FPS
567 frames in 8.7 seconds = 65.423 FPS
531 frames in 8.7 seconds = 61.269 FPS
^C
```

I saw a window with 3 gears in it.


```
[dharper@micc ~]$ glxinfo 
name of display: localhost:17.0
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
 Major opcode of failed request:  149 (GLX)
 Minor opcode of failed request:  6 (X_GLXIsDirect)
 Serial number of failed request:  31
 Current serial number in output stream:  30
```

>>>

**Comment by pauldmccarthy** [comment]
Weird that `glxinfo` crashes, but `glxgears` doesn't. In any case, this suggests a problem with the environment, rather than an issue with FSLeyes - if `glxinfo` isn't working, then FSLeyes won't work.

 I'd suggest playing with the `LIBGL_ALWAYS_INDIRECT`  variable, e.g.:

```
export LIBGL_ALWAYS_INDIRECT=1
glxinfo 
# and if glxinfo works
fsleyes
```

[This page](https://www.scm.com/doc/Installation/Remote_GUI.html#opengl-direct-or-indirect-rendering) contains a lot of useful information on running OpenGL remotely

**Comment by dmd** [comment]
I'm stumped, you?

**Comment by pauldmccarthy** [comment]
is the remote machine powerful? You could try using `LIBGL_ALWAYS_SOFTWARE=1` to force software-based rendering. Also worth upgrading to the latest XQuartz (and double checking that GLX is enabled)

**Comment by dmd** [comment]
Totally stumped. Doing `export LIBGL_ALWAYS_SOFTWARE=1` had no effect. Reinstalled XQuartz 2.8.1, made sure `defaults read org.xquartz.X11 enable_iglx` returns `1`, same thing.

**Comment by pauldmccarthy** [comment]
I'm afraid I don't know either ... unless this is a Wayland/EGL issue? What version of wxPython is installed? 

wxpython 4.1.0 and newer default to using EGL, rather than GLX, for interacting with OpenGL. I'm not sure of how things work when running over SSH, and don't have a mac to play with, but perhaps you might have success by setting the `GDK_BACKEND=x11` variable, or by down-grading wxpython to 4.0.7.post2

**Comment by dmd** [comment]
Seems like we're already down there:

```
$ /cm/shared/fsl-6.0.4/fslpython/envs/fslpython/bin/python
Python 3.7.6 | packaged by conda-forge | (default, Jun  1 2020, 18:57:50)
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import wx
>>> wx.version()
'4.0.7.post2 gtk2 (phoenix) wxWidgets 3.0.5'
```

**Comment by dmd** [comment]
And again, this is just *one user*. Everyone else - using the same version of XQuartz, and the same version of MacOS, is doing fine, so ... i have no idea.

**Comment by pauldmccarthy** [comment]
Is there something different about this user's environment? E.g. do they have a different conda or virtualenv environment active? Could you ask them to run `env` and send you the output?

**Comment by dmd** [comment]
It's almost definitely their Mac's fault; if I log in as them from MY Mac, everything works fine.

**Comment by pauldmccarthy** [comment]
Perhaps it's worth purging XQuartz from their system, then re-installing?

1. Purge XQuartz via the instructions under *Uninstall (XQuartz 2.8 and later)*: https://www.xquartz.org/FAQs.html
2. Reboot (probably not necessary, but never hurts)
3. Re-install XQuartz 2.8.1
4. Run `defaults write org.xquartz.X11 enable_iglx -bool true`
5. SSH to remote server, and try running `glxinfo` / `fsleyes`

**Comment by sharrison5** [comment]
I found this thread really useful when I ran into what I assume is the same issue – thanks! I was seeing similar weirdness with `glxinfo` and `glxgears`, so again something on my Mac seems the likely culprit: possibly originating from https://github.com/XQuartz/XQuartz/issues/144? Ultimately, I had to roll back to 2.7.11 as reinstalling / rebooting didn't fix either 2.8.1 or 2.8.0.

**Comment by pauldmccarthy** [comment]
@sharrison5 Aah thanks - I hadn't come across that XQuartz issue - so perhaps it's a regression in XQuartz 2.8.0 ...
